[git-webkit] Support reverting multiple commits at once

briannafan opened the pull request:
#### 627b31c8383b01ea97f1b4d4cd95e28e73467dc0
<pre>
[git-webkit] Support reverting multiple commits at once
<a href="https://bugs.webkit.org/show_bug.cgi?id=239796">https://bugs.webkit.org/show_bug.cgi?id=239796</a>
<a href="https://rdar.apple.com/problem/92702616">rdar://problem/92702616</a>

Reviewed by Jonathan Bedard.

Allows for multiple commits to be passed in as args.
Includes all reverted commit identifiers in the revert message.
Fixes bug in issue creation, adds support for radar importer.

* Tools/Scripts/hooks/prepare-commit-msg:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert)
(Revert.parser):
(Revert.get_issue_info):
(Revert.create_revert_commit_msg):
(Revert.revert_commit):
(Revert.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
(TestRevert.test_github):
(TestRevert.test_github_two_step):
(TestRevert.test_args):
(test_update):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch):
(Branch.create_radar):
(Branch.main):

Canonical link: <a href="https://commits.webkit.org/271587@main">https://commits.webkit.org/271587@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67cf58dde127e3a5b15c5e41031fc2ed8ed06600

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30213 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31459 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26309 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29412 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9618 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4843 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26355 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29113 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6197 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24778 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5389 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29000 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5535 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25788 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32797 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26393 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26230 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31776 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5509 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3677 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29557 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/28609 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7136 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6904 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5986 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6030 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -94,65 +94,73 @@ def get_issue_info(cls, args, repository, **kwargs):
     @classmethod
-    def revert_commit(cls, args, repository, issue, commit, commit_title, bug_urls, **kwargs):
-        result = run([repository.executable(), 'revert', '--no-commit'] + [commit.hash], cwd=repository.root_path, capture_output=True)
+    def revert_commit(cls, args, repository, issue, commit_identifiers, commits_to_revert, **kwargs):
```
If you just passed `commits`, you could print commit objects. This is probably best anyways, while WebKit uses identifiers, this script should work in repositories which don't.

JonWBedard reviewed (commented):
I think the biggest question I have about this PR: do we _want_ to handle multi-commit reverts as a single revert, or as each commit reverted in sequence? We reverted all commits in a single commit during Subversion days because we couldn't atomically push multiple commits in a single operation. We _can_ do that in `git`, though. And there is precedence for multi-commit PRs for cherry-picks, this seems conceptually similar.

I would like to discuss this with @briannafan and a few other folks. There are good arguments for both approaches, I want to make sure we're transparently evaluating both techniques.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -99,6 +99,35 @@ def to_branch_name(cls, value):
         return string_utils.encode(result, target_type=str)
 
+    @classmethod
+    def create_radar(cls, args, repository, issue):
```
Should name this `cc_radar` or even ~`cced_radar`~ to indicate it will _return_ the cced radar.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -87,72 +94,95 @@ def get_issue_info(cls, args, repository, **kwargs):
+            commits_to_revert.append(commit.hash)
+            bug_urls = []
+            commit_title = None
+            for line in commit.message.splitlines():
```
Can we just use `commit.issues` here? Notably, this won't work in repositories which do `<bug> [title]`

briannafan committed:
[git-webkit] Support reverting multiple commits at once
https://bugs.webkit.org/show_bug.cgi?id=239796
rdar://problem/92702616

Reviewed by Jonathan Bedard.

Allows for multiple commits to be passed in as args.
Includes all reverted commit identifiers in the revert message.
Fixes bug in issue creation, adds support for radar importer.

* Tools/Scripts/hooks/prepare-commit-msg:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert)
(Revert.parser):
(Revert.get_issue_info):
(Revert.create_revert_commit_msg):
(Revert.revert_commit):
(Revert.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
(TestRevert.test_github):
(TestRevert.test_github_two_step):
(TestRevert.test_args):
(test_update):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch):
(Branch.create_radar):
(Branch.main):

Canonical link: https://commits.webkit.org/271587@main
